contracts/fuses/napier/NapierCombineFuse.sol

```diff
@@ -0,0 +1,99 @@
+struct NapierCombineFuseEnterData {
+    IPrincipalToken principalToken;
+    address tokenOut;
+    uint256 principals;
+}
+
+/// @title NapierCombineFuse
+/// @notice Fuse for supplying assets into Napier V2 Principal Tokens via the universal router
+/// @dev Substrates in this fuse are the Napier V2 Principal Tokens
+contract NapierCombineFuse is NapierUniversalRouterFuse {
+    using SafeERC20 for ERC20;
+
+    /// @notice Emitted when supplying assets into Napier V2 PTs
+    /// @param version Address of this contract version
+    /// @param principalToken Address of the Napier V2 Principal Token
+    /// @param tokenOut Asset supplied into the router
+    /// @param amountOut Amount of tokens received
+    event NapierCombineFuseEnter(address version, address principalToken, address tokenOut, uint256 amountOut);
+
+    constructor(uint256 marketId_, address router_) {
+        VERSION = address(this);
+        if (marketId_ == 0) revert NapierFuseIInvalidMarketId();
+        if (router_ == address(0)) revert NapierFuseIInvalidRouter();
+
+        MARKET_ID = marketId_;
+        ROUTER = IUniversalRouter(router_);
+    }
+
+    /// @notice Early redeem PT and YT into the tokenOut (before/after the maturity)
+    function enter(NapierCombineFuseEnterData calldata data_) external {
+        IPrincipalToken pt = data_.principalToken;
+
+        if (!PlasmaVaultConfigLib.isSubstrateAsAssetGranted(MARKET_ID, address(data_.principalToken))) {
+            revert NapierFuseIInvalidMarketId();
+        }
+
+        address yt = pt.i_yt();
+        if (!PlasmaVaultConfigLib.isSubstrateAsAssetGranted(MARKET_ID, yt)) {
+            revert NapierFuseIInvalidMarketId();
+        }
+        if (data_.principals == 0) {
+            return;
+        }
+
+        address underlying = pt.underlying();
+        address asset = pt.i_asset();
+
+        bytes memory commands;
+        bytes[] memory inputs;
+        if (data_.tokenOut == underlying) {
+            // Combine PT and YT for underlying token
+            commands = abi.encodePacked(bytes1(uint8(Commands.PT_COMBINE)));
+            inputs = new bytes[](1);
+            inputs[0] = abi.encode(pt, ActionConstants.CONTRACT_BALANCE, address(this));
+        } else if (data_.tokenOut == asset) {
+            // Combine PT and YT for tokenOut through VaultConnector
+            commands = abi.encodePacked(
+                bytes1(uint8(Commands.PT_COMBINE)),
+                bytes1(uint8(Commands.VAULT_CONNECTOR_REDEEM))
+            );
+
+            inputs = new bytes[](2);
+            inputs[0] = abi.encode(pt, ActionConstants.CONTRACT_BALANCE, ActionConstants.ADDRESS_THIS);
+            inputs[1] = abi.encode(underlying, asset, data_.tokenOut, ActionConstants.CONTRACT_BALANCE, address(this));
+        } else {
+            revert NapierFuseIInvalidToken();
+        }
+
+        uint256 balanceBefore = ERC20(data_.tokenOut).balanceOf(address(this));
```

Comment:
All addresses that come in the parameters and on which actions are performed should be in substrata for security reasons.